aten/src/ATen/native/TensorCompare.cpp

```diff
@@ -355,15 +377,30 @@ std::vector<Tensor> where(const Tensor& condition) {
 }
 
 Tensor _s_where(const Tensor& condition, const Tensor& self, const Tensor& other) {
-  TORCH_CHECK(self.dtype() == other.dtype(), "expected scalar type ", self.dtype(), " but found ", other.dtype());
-  Tensor ret = at::empty(self.sizes(), self.options());
+  // Here we do the type promotion without the TensorIterator machinery.
+  // For `where` we only want to consider `self` and `other`,
+  // however right now TensorIterator API doesn't have a way to specify which inputs
+  // should take part in type promotion.
+  auto common_dtype = at::result_type(self, other);
+  Tensor ret = at::empty(self.sizes(), self.options().dtype(common_dtype));
+
+  auto get_promoted = [&common_dtype](const Tensor& t) {
+    // dynamic casting on CPU is not allowed
```

Comment:
This is interesting. Am I understanding correctly that it's due to https://github.com/pytorch/pytorch/pull/28427#issuecomment-545687342?

Reply:
That looks like the motivation. 
There is a check in `cpu_kernel` for the same (which fails without this casting here)
https://github.com/pytorch/pytorch/blob/30a7c768d77d870c6194f5b6cb0bbd6fe1211069/aten/src/ATen/native/cpu/Loops.h#L257-L258
(called from here)
https://github.com/pytorch/pytorch/blob/30a7c768d77d870c6194f5b6cb0bbd6fe1211069/aten/src/ATen/native/cpu/TensorCompareKernel.cpp#L208

AFAIK, TensorIterator does similar things for CPU and type promotion

cc: @mruberry